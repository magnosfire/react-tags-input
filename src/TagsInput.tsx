import React, { useCallback, useEffect, useState } from "react";
import Chip from "@material-ui/core/Chip";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/core";
import validator from 'validator';

export default function TagsInput(props) {

  const classes = useStyles();
  const { onSelectTags, tags, ...other } = props;
  const [selectedItem, setSelectedItem] = useState([]);
  const [newTag, setNewTag] = useState('');
  const [, updateState] = useState(0);

  useEffect(() => {

    setSelectedItem(tags);

  }, [tags]);

  const handleDelete = (item: string) => () => {

    const index = tags.indexOf(item);

    if (index > -1) {
      tags.splice(index, 1);
    }

    updateState(Math.random());
    setSelectedItem(tags);

    return newTag;

  };

  const handleDeleteLastItem = () => {

    tags.pop();

    updateState(Math.random());
    setSelectedItem(tags);

    return tags;

  };

  const handleNewTag = (event) => {

    setNewTag(event.target.value);
    return event.target.value;

  };

  const handleTag = () => {

    if (!tags.includes(newTag)) {

      if (newTag.includes(';')) {

        const tagsArray: string[] = newTag.split(';');

        tagsArray.map((email) => {
          if (validator.isEmail(email)) {
            tags.push(email);
          }
        });

      } else {

        if (validator.isEmail(newTag)) {
          tags.push(newTag);
        }

      }

    }

    setNewTag('');

  }

  const handleKeyPressed = (event) => {

    switch (event.key) {
      case 'Backspace': {
        if (newTag.length == 0) {
          handleDeleteLastItem();
        }
        break;
      }
      case 'Enter': {
        handleTag();
        break;
      }
      case 'Tab': {
        event.preventDefault()
        handleTag();
        break;
      }
      default:
        break;
    }

  }

  return (
    <TextField
      value={newTag}
      onChange={handleNewTag}
      onKeyDown={(event) => handleKeyPressed(event)}
      className={classes.textField}
      InputProps={{
        startAdornment: selectedItem.map((item) => (
          <Chip
            key={item}
            tabIndex={-1}
            label={item}
            variant={"outlined"}
            color={'primary'}
            className={classes.chip}
            onDelete={handleDelete(item)}
          />
        )),
        classes: {
          root: classes.root
        }
      }}
      {...other}
    />
  );
}

TagsInput.defaultProps = {
  tags: []
};

const useStyles = makeStyles(() => ({
  root: {
    flexWrap: "wrap",
    paddingTop: 16
  },
  chip: {
    marginRight: 8,
    marginBottom: 8,
    fontWeight: 700,
  },
  textField: {
    borderBottomColor: "red",
    borderBottom: 1
  }
}));