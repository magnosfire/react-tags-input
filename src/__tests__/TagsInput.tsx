import React from 'react'
import TagsInput from '@/TagsInput'
import { cleanup, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

describe('TagsInput Component', () => {
  afterEach(jest.clearAllMocks)
  afterEach(cleanup)

  it('cria o snapshot do component', () => {
    const container = render(<TagsInput />)
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('deve renderizar as tags enviadas por atributos', () => {
    const emails = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']
    const { debug } = render(<TagsInput tags={emails} />)

    debug()
  })

  it('deve renderizar tags quando preencher o input e pressionar enter', () => {

    const emails = [];
    const container = render(<TagsInput inputProps={{ 'aria-label': 'tag' }} tags={emails} />)
    const input = container.getByRole('textbox', {name: /tag/i})

    userEvent.type(input, 'hire-me@email.com{enter}')

    const { debug } = container;

    debug()

  })

  it('deve renderizar tags quando preencher o input e pressionar tab', () => {
  
    const emails = [];
    const container = render(<TagsInput inputProps={{ 'aria-label': 'tag' }} tags={emails} />)
    const input = container.getByRole('textbox', {name: /tag/i});

    userEvent.type(input, 'hire-me@email.com{enter}')

    const { debug } = container;

    debug()


  })

  it('deve deletar a útima tag criada ao pressionar o botão de backspace', async () => {

    const emails = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']
    const container = render(<TagsInput inputProps={{ 'aria-label': 'tag' }} tags={emails} />)
    const input = container.getByRole('textbox', {name: /tag/i});

    userEvent.type(input, '{backspace}')

    const { debug } = container;

    debug()

    expect(emails == ['contato@rarolabs.com.br'])
  })
})
