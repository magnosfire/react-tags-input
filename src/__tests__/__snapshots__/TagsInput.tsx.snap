// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsInput Component cria o snapshot do component 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-textField-3"
  >
    <div
      class="MuiInputBase-root MuiInput-root makeStyles-root-1 MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
